resources/net.go

```diff
@@ -0,0 +1,804 @@
+
+	// opState is the name of the dbus property for the network interface's
+	// "operational state." It is used and to check the status of the network
+	// connection.
+	opState = "OperationalState"
+	// opStateRoutable us the routable operational state.
+	opStateRoutable = "routable"
+	// opStateNoCarrier is the no-carrier operational state.
+	opStateNoCarrier = "no-carrier"
+	// opStateOff is the off operational state.
+	opStateOff = "off"
+
+	// adminState is the name of the dbus property for the network interface's
+	// "administrative state." It is used and to check if the interface is
+	// configured or in a transitional state.
+	adminState = "AdministrativeState"
+	// adminStateConfiguring is the administrative state of the interface when
+	// it is in the process of configuring.
+	adminStateConfiguring = "configuring"
+
+	// eventSources is the number of event sources in watch.
+	eventSources = 5
+)
+
+func init() {
+	RegisterResource("net", func() Res { return &NetRes{} })
+}
+
+// NetRes is an systemd-networkd resource.
+type NetRes struct {
+	BaseRes `yaml:",inline"`
+	State   string   `yaml:"state"`   // up or down or empty
+	DHCP    bool     `yaml:"dhcp"`    // enable dhcp
+	Addrs   []string `yaml:"addrs"`   // list of addresses in cidr format
+	Gateway string   `yaml:"gateway"` // gateway address
+
+	ifacePath    dbus.ObjectPath // the dbus path to the interface
+	unitFilePath string          // the interface unit file path
+
+	// internal busses
+	sdbus *systemd.Conn // go-systemd dbus connection
+	gdbus *dbus.Conn    // godbus dbus connection
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *NetRes) Default() Res {
+	return &NetRes{
+		BaseRes: BaseRes{
+			MetaParams: DefaultMetaParams, // force a default
+		},
+		State: ifaceUp,
+	}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *NetRes) Validate() error {
+	// validate state
+	if obj.State != ifaceUp && obj.State != ifaceDown && obj.State != "" {
+		return fmt.Errorf("invalid state: %s", obj.State)
+	}
+
+	// validate dhcp
+	if obj.DHCP && obj.Addrs != nil {
+		return fmt.Errorf("cannot assign IP with DHCP enabled")
+	}
+
+	// validate network address input
+	if obj.Addrs != nil {
+		for _, addr := range obj.Addrs {
+			if _, _, err := net.ParseCIDR(addr); err != nil {
+				return errwrap.Wrapf(err, "error parsing address: %s", addr)
+			}
+		}
+	}
+
+	if obj.Gateway != "" {
+		if g := net.ParseIP(obj.Gateway); g == nil {
+			return fmt.Errorf("could not parse gateway: %s", obj.Gateway)
+		}
+	}
+
+	// validate the interface name
+	_, err := net.InterfaceByName(obj.GetName())
+	if err != nil {
+		return errwrap.Wrapf(err, "error finding interface: %s", obj.GetName())
+	}
+
+	return obj.BaseRes.Validate()
+}
+
+// Init runs some startup code for this resource.
+func (obj *NetRes) Init() error {
+	var err error
+	// get the network link's dbus path
+	obj.ifacePath, err = dbusIfacePath(obj.GetName())
+	if err != nil {
```

Comment:
nit, but fwiw when you're doing an `=` instead of an `:=` you can prefix this, eg:

```
if obj.ifacePath, err = dbusIfacePath(obj.GetName()); err != nil {
 ...
}

```

In general I like to do those as much as possible, unless the line is extremely long.